File: sources/pascal/03_func_sum/main.pas
{ task 03 func_sum — expected output: 100000000 }
{ build: fpc -O3 -oprogram main.pas    run: ./program }
{ Windows x64: i386-win32 fpc + cross.x86_64-win64 add-on, build with -Px86_64; run as program.exe }

program prog;
{$mode objfpc}{$H+}

uses
  SysUtils;

{ fpc only inlines a routine that carries the "inline" modifier, and only while the
  $INLINE switch is on; both are off here, so the hundred million calls really happen. }
{$INLINE OFF}

function add_one(n: Int64): Int64;
{$INLINE OFF}
begin
  Result := n + 1;
end;

var
  i, value: Int64;
begin
  value := 0;
  for i := 1 to 100000000 do
    value := add_one(value);
  WriteLn(value);
end.

File: sources/fortran/03_func_sum/main.f90
! task 03 func_sum — expected output: 100000000
! build: gfortran -O3 -o prog main.f90 add_one.f90 (flang -O3 -o prog main.f90 add_one.f90)    run: ./prog
!
! add_one sits in its own file: cross-file inlining needs link-time
! optimization, which is off here, so the hundred million calls really happen.

program main
  implicit none
  integer(kind=8) :: value, i
  integer(kind=8), external :: add_one

  value = 0_8
  do i = 1_8, 100000000_8
    value = add_one(value)
  end do

  write(*,'(i0)') value
end program main
